src/Microsoft.TestPlatform.Build/Tasks/VSTestTask.cs

```diff
@@ -268,47 +275,57 @@ private List<string> AddArgs()
                 }
             }
 
-            if (isCollectCodeCoverageEnabled || isRunSettingsEnabled)
-            {
-                // Pass TraceDataCollector path to vstest.console as TestAdapterPath if --collect "Code Coverage"
-                // or --settings (User can enable code coverage from runsettings) option given.
-                // Not parsing the runsettings for two reason:
-                //    1. To keep no knowledge of runsettings structure in VSTestTask.
-                //    2. Impact of adding adapter path always is minimal. (worst case: loads additional data collector assembly in datacollector process.)
-                // This is required due to currently trace datacollector not ships with dotnet sdk, can be remove once we have
-                // go code coverage x-plat.
-                if (!string.IsNullOrEmpty(this.VSTestTraceDataCollectorDirectoryPath))
-                {
-                    allArgs.Add("--testAdapterPath:" +
-                                ArgumentEscaper.HandleEscapeSequenceInArgForProcessStart(this
-                                    .VSTestTraceDataCollectorDirectoryPath));
-                }
-                else
-                {
-                    if (isCollectCodeCoverageEnabled)
-                    {
-                        // Not showing message in runsettings scenario, because we are not sure that code coverage is enabled.
-                        // User might be using older Microsoft.NET.Test.Sdk which don't have CodeCoverage infra.
-                        Console.WriteLine(Resources.UpdateTestSdkForCollectingCodeCoverage);
-                    }
-                }
-            }
+            this.AddCodeCoveragePathInTestAdapter(isCollectCodeCoverageEnabled, isRunSettingsEnabled, this.VSTestTraceDataCollectorDirectoryPath, allArgs, Resources.UpdateTestSdkForCollectingCodeCoverage);
+            this.AddCodeCoveragePathInTestAdapter(isXPlatCollectCodeCoverageEnabled, isRunSettingsEnabled, this.VSTestXPlatDataCollectorDirectoryPath, allArgs, Resources.UpdateTemplateForCollectingXPlatCodeCoverage);
 
-            if(!string.IsNullOrWhiteSpace(this.VSTestNoLogo))
+            if (!string.IsNullOrWhiteSpace(this.VSTestNoLogo))
             {
                 allArgs.Add("--nologo");
             }
 
             return allArgs;
         }
+
+        /// <summary>
+        /// Adds code coverage path in test adapter.
+        /// </summary>
+        /// <param name="isCollectCodeCoverageEnabled"></param>
+        /// <param name="isRunSettingsEnabled"></param>
+        /// <param name="adapterPath"></param>
+        /// <param name="allArgs"></param>
+        /// <param name="warningMessage"></param>
+        private void AddCodeCoveragePathInTestAdapter(bool isCollectCodeCoverageEnabled, bool isRunSettingsEnabled, string adapterPath, List<string> allArgs, string warningMessage)
```

Comment:
The job is this method should simply to add a given adapter path to arguments. It doesn't need to check for runsettings, & coverage. Make these checks before invoking

Reply:
Done. Would need to send if coverage is enabled or not for showing the warning message specific to `code coverage` or `xplat code coverage`